### PR TITLE
[MkFit] Use binnor for seed and hit sorting, remove Ice/

### PR DESCRIPTION
**Pull request by osschar**
### Changes
- Add radix-sort option in binnor.
- While importing seeds use binnor for sorting or rely on previous sorting done during seed cleaning.
- Split track-related classes out of HitStructures.h/cc into a new TrackStructures.h/cc file-set.
- Use binnor with radix_sort in LayerOfHits.
- Use single std::vector for storing dead-bin information.
- Remove MkFitCore/src/Ice/ directory, remove `-Wno-error=strict-aliasing` from  
RecoTracker/MkFitCore/BuildFile.xml.
- Improve std-vs-radix sort heuristic, cleanup binnor_demo.cxx (use same axis parameters as in LayerOfHits; perform timing-loop multiple times).

### Validation
MTV plots http://uaf-10.t2.ucsd.edu/~mmasciov/MIC/IceOut/MTV_mkFit_TTbarPU_from1240pre2_cmssw1240pre2_IceOut/

Efficiencies remain unchanged. There are minor fluctuations in fake/duplicate rates induced by more precise sorting of hits in phi and potentially also from additional sorting of incoming seeds in phi (they were only sorted in eta before). There are additional hit-on-track processing changes expected in treatment of dead-regions, large clusters, and through usage of the additional per-module information that is now available -- following all these selection windows will be retuned to achieve optimal performance.

### Related issues
- This PR addresses the second long term issue in #36966.
- Also included are minor log-printout changes requested by @qliphy in #37418 that were postponed in that PR (see commit 958ba5f).

### CODE REVIEW

File: RecoTracker/MkFitCore/src/MkFinder.cc [hunk 1]
```diff
@@ -584,85 +576,44 @@ namespace mkfit {
...
+                         st_pt, st_eta, st_phi,
+                         st_nhits, st_charge, st_r, st_z,
+                         q, L.hit_q(hi), ddq, dq, phi, L.hit_phi(hi), ddphi, dphi,
+                         tx, ty, tr, tphi, tz,
+                         tex, tey, ter, tephi, tez,
+                         hx, hy, hr, hphi, hz,
+                         hex, hey, her, hephi, hez,
+                         ht_dxy, ht_dz, ht_dphi,
+                         hchi2);
+                  // clang-format ofn
```
Comment (slava77) [inline RecoTracker/MkFitCore/src/MkFinder.cc:614]: typo?

Comment (osschar) [reply RecoTracker/MkFitCore/src/MkFinder.cc]: Yup, one backspace push too little, maybe I should oil it. Fixed.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
given the amount of new code, I would also ask tracking to take a look in the subsequent comment.

**Comment by jpata** [comment]
assign @cms-sw/tracking-pog-l2

**Comment by osschar** [comment]
> given the amount of new code, I would also ask tracking to take a look in the subsequent comment.

There isn't really a lot of new code ... but I have moved track-candidate related structures out of HitStructures.h/cc into new files TrackStructures.h/cc to make things more manageable (and to reduce "bundling" and over-inclusion).

**Comment by mmusich** [comment]
type tracking

**Comment by slava77** [comment]
@cms-sw/tracking-pog-l2 @mmusich @cms-sw/reconstruction-l2 
please clarify on the plans for your review of this PR.
Please expedite if possible.
We have more PRs in mkFIt that would depend on this merged.
Thank you.

**Comment by mmusich** [comment]
+1
   * technical
   * addresses the second long term issue in https://github.com/cms-sw/cmssw/issues/36966.

**Comment by jpata** [comment]
+reconstruction
- improved hit sorting and code reorganization
- minor changes to tracks and everything downstream (per the usual reco language, not a technical PR)

**Comment by perrotta** [comment]
Is there any evaluation of the effect on timing? Modifying sorting I'd expect can have some...

**Comment by slava77** [comment]
test parameters:
- workflow_profiling = 11834.21
- enable_test = profiling

**Comment by slava77** [comment]
@smuzaffar 
is there a way to run only a profiling test (after the build), skip the relvals and unit tests?

**Comment by smuzaffar** [comment]
@slava77 , no sorry currently there is no way to do this.

**Comment by osschar** [comment]
> Is there any evaluation of the effect on timing? Modifying sorting I'd expect can have some...

I actually ported radix sort of integers into binnor so I don't expect significant changes (if any, I guess they will be in the noise). But yes, makes sense to check :)

**Comment by slava77** [comment]
> **Additional Tests**: PROFILING

- `MkFitProducer` igprof from [24.34](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-26-1100/slc7_amd64_gcc10/profiling/11834.21/igprofCPU_step3.19/32) to [24.74](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-26-1100/slc7_amd64_gcc10/profiling/11834.21/PR-3a9357/24228/igprofCPU_step3.19/31)
    - essentially all of it is in `mkfit::run_OneIteration` from [24.09](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-26-1100/slc7_amd64_gcc10/profiling/11834.21/igprofCPU_step3.19/35) to [24.57](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-26-1100/slc7_amd64_gcc10/profiling/11834.21/PR-3a9357/24228/igprofCPU_step3.19/34) . Most of the difference seems accidental though, either from timing reproducibility or from a change in order of some objects.
        - `find_tracks_load_seeds` is the affected function call, its cost goes down 1.67 to 1.56 
- `MkFitEventOfHitsProducer` from [0.51](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-26-1100/slc7_amd64_gcc10/profiling/11834.21/igprofCPU_step3.19/1574) to [0.61](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-26-1100/slc7_amd64_gcc10/profiling/11834.21/PR-3a9357/24228/igprofCPU_step3.19/1351)
    - `mkfit::LayerOfHits::endRegistrationOfHits` from [0.09](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-26-1100/slc7_amd64_gcc10/profiling/11834.21/igprofCPU_step3.19/3673) to [0.15](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_4_X_2022-04-26-1100/slc7_amd64_gcc10/profiling/11834.21/PR-3a9357/24228/igprofCPU_step3.19/2956) this is probably the clearest measure (not mixed with other calls)
- overall, there seems to be up to  0.5 (0.1% of the job cost) increase in cost, but the more likely value is closer to 0.1 (0.02%) based on the analysis of related contributions.

**Comment by perrotta** [comment]
+1
- Thank you @slava77, all [this](https://github.com/cms-sw/cmssw/pull/37586#issuecomment-1110142151) looks quite reasonable

**Pull request merged**